apps/gdalalg_vector_pipeline.cpp

```diff
@@ -0,0 +1,612 @@
+            {
+                writeAlg.m_inputDataset.Set(m_outputDataset.GetDatasetRef());
+                if (writeAlg.Run(pfnProgress, pProgressData))
+                {
+                    m_outputDataset.Set(
+                        writeAlg.m_outputDataset.GetDatasetRef());
+                    ret = true;
+                }
+            }
+        }
+
+        return ret;
+    }
+    else
+    {
+        return RunStep(pfnProgress, pProgressData);
+    }
+}
+
+/************************************************************************/
+/*        GDALVectorPipelineAlgorithm::GDALVectorPipelineAlgorithm()    */
+/************************************************************************/
+
+GDALVectorPipelineAlgorithm::GDALVectorPipelineAlgorithm()
+    : GDALVectorPipelineStepAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      /*standaloneStep=*/false)
+{
+    AddInputArgs(/* hiddenForCLI = */ true);
+    AddProgressArg();
+    AddArg("pipeline", 0, _("Pipeline string"), &m_pipeline)
+        .SetHiddenForCLI()
+        .SetPositional();
+    AddOutputArgs(/* hiddenForCLI = */ true,
+                  /* shortNameOutputLayerAllowed=*/false);
+
+    m_stepRegistry.Register<GDALVectorReadAlgorithm>();
+    m_stepRegistry.Register<GDALVectorWriteAlgorithm>();
+    m_stepRegistry.Register<GDALVectorReprojectAlgorithm>();
+    m_stepRegistry.Register<GDALVectorFilterAlgorithm>();
+}
+
+/************************************************************************/
+/*              GDALVectorPipelineAlgorithm::GetStepAlg()               */
+/************************************************************************/
+
+std::unique_ptr<GDALVectorPipelineStepAlgorithm>
+GDALVectorPipelineAlgorithm::GetStepAlg(const std::string &name) const
+{
+    auto alg = m_stepRegistry.Instantiate(name);
+    return std::unique_ptr<GDALVectorPipelineStepAlgorithm>(
+        cpl::down_cast<GDALVectorPipelineStepAlgorithm *>(alg.release()));
+}
+
+/************************************************************************/
+/*       GDALVectorPipelineAlgorithm::ParseCommandLineArguments()       */
+/************************************************************************/
+
+bool GDALVectorPipelineAlgorithm::ParseCommandLineArguments(
+    const std::vector<std::string> &args)
+{
+    if (args.size() == 1 && (args[0] == "-h" || args[0] == "--help" ||
+                             args[0] == "help" || args[0] == "--json-usage"))
+        return GDALAlgorithm::ParseCommandLineArguments(args);
+
+    for (const auto &arg : args)
+    {
+        if (arg.find("--pipeline") == 0)
+            return GDALAlgorithm::ParseCommandLineArguments(args);
+
+        // gdal vector pipeline [--progress] "read poly.gpkg ..."
+        if (arg.find("read ") == 0)
+            return GDALAlgorithm::ParseCommandLineArguments(args);
+    }
+
+    if (!m_steps.empty())
+    {
+        ReportError(CE_Failure, CPLE_AppDefined,
```

Comment:
This error message can only happen when (mis)using the C/C++ API and invoking that method twice. Not with command line utilities. I'm not sure how it could be make clearer. Suggestions?

Reply:
If it indicates a programming error, I think I would just assert.

Reply:
I'm not a big fan of assert() that aren't related to checking "impossible conditions" / invariants in the code base itself. Here, it is a situation that could be triggered from Python, and we don't want Python to crash. Users will get a clean Python exception, which is much better. And assert() are optimized out in release builds.